build.go

```diff
@@ -0,0 +1,161 @@
+// containing the Bill-of-Materials of a given dependency.
+type BuildPlanRefinery interface {
+	BillOfMaterial(dependency postal.Dependency) packit.BuildpackPlan
+}
+
+// InstallProcess defines the interface for installing the pip dependency into a layer.
+type InstallProcess interface {
+	Execute(srcPath, targetLayerPath string) error
+}
+
+// Build will return a packit.BuildFunc that will be invoked during the build
+// phase of the buildpack lifecycle.
+//
+// Build will find the right pip dependency to install, install it in a
+// layer, and generate Bill-of-Materials. It also makes use of the checksum of
+// the dependency to reuse the layer when possible.
+func Build(installProcess InstallProcess, entries EntryResolver, dependencies DependencyManager, planRefinery BuildPlanRefinery, logs scribe.Emitter, clock chronos.Clock) packit.BuildFunc {
+	return func(context packit.BuildContext) (packit.BuildResult, error) {
+		logs.Title("%s %s", context.BuildpackInfo.Name, context.BuildpackInfo.Version)
+
+		logs.Process("Resolving Pip version")
+
+		entry, sortedEntries := entries.Resolve(Pip, context.Plan.Entries, Priorities)
+
+		logs.Candidates(sortedEntries)
+
+		version, _ := entry.Metadata["version"].(string)
+
+		dependency, err := dependencies.Resolve(filepath.Join(context.CNBPath, "buildpack.toml"), entry.Name, version, context.Stack)
+		if err != nil {
+			return packit.BuildResult{}, err
+		}
+
+		dependency.Name = "Pip"
+		logs.SelectedDependency(entry, dependency, clock.Now())
+
+		bom := planRefinery.BillOfMaterial(dependency)
+
+		pipLayer, err := context.Layers.Get(Pip)
+		if err != nil {
+			return packit.BuildResult{}, err
+		}
+
+		cachedSHA, ok := pipLayer.Metadata[DependencySHAKey].(string)
+		if ok && cachedSHA == dependency.SHA256 {
+			logs.Process("Reusing cached layer %s", pipLayer.Path)
+			return packit.BuildResult{
+				Plan:   bom,
+				Layers: []packit.Layer{pipLayer},
+			}, nil
+		}
+
+		pipLayer, err = pipLayer.Reset()
+		if err != nil {
+			return packit.BuildResult{}, err
+		}
+
+		pipLayer.Launch, pipLayer.Build = entries.MergeLayerTypes(Pip, context.Plan.Entries)
+		pipLayer.Cache = pipLayer.Build
+
+		pipSrcDir, err := ioutil.TempDir("", "pip-source")
+		if err != nil {
+			return packit.BuildResult{}, fmt.Errorf("failed to create temp pip-source dir: %w", err)
+		}
+
+		logs.Process("Executing build process")
+		logs.Subprocess(fmt.Sprintf("Installing Pip %s", dependency.Version))
+
+		duration, err := clock.Measure(func() error {
+			err = dependencies.Install(dependency, context.CNBPath, pipSrcDir)
+			if err != nil {
+				return err
+			}
+			return installProcess.Execute(pipSrcDir, pipLayer.Path)
+		})
+		if err != nil {
+			return packit.BuildResult{}, err
+		}
+
+		logs.Action("Completed in %s", duration.Round(time.Millisecond))
+		logs.Break()
+
+		var sitePackagesPath string
+		err = filepath.Walk(pipLayer.Path, func(path string, info os.FileInfo, err error) error {
```

Comment:
I'm not really familiar with `site-packages` - is there a reason we're traversing the filesystem instead of running `python -m site --user-site` or are they not the same?

Reply:
In full honesty, I wasn't award of the `--user-site` flag when we implemented this code. I will try it out and see. There is no special reason for traversing the file system , we really just need the full path to the site packages directory so we can add it onto the PYTHONPATH.

Reply:
I think we should do it the right way and not check for an arbitrary file/dir named site-packages. It can be extracted out as a site-process for example.
I tried:
```diff
                if sitePackagesPath == "" {
                        return packit.BuildResult{}, fmt.Errorf("pip installation failed: site packages are missing from the pip layer")
                }
+               logs.Process("sitePackages=%s", sitePackagesPath)
+               cmd := exec.Command("python", "-m", "site", "--user-site")
+               var out bytes.Buffer
+               cmd.Stdout = &out
+               cmd.Env = os.Environ()
+               cmd.Env = append(cmd.Env, fmt.Sprintf("PYTHONUSERBASE=%s", pipLayer.Path))
+               err = cmd.Run()
+               if err != nil {
+                       return packit.BuildResult{}, fmt.Errorf("site-command-error: %w", err)
+               }
+               logs.Process("sitecommandresponse=%s", out.String())
+
                pipLayer.SharedEnv.Prepend("PYTHONPATH", sitePackagesPath, ":")
```
and got the same value as it's done now:
```
[builder]   sitePackages=/layers/paketo-community_pip/pip/lib/python3.8/site-packages
[builder]   sitecommandresponse=/layers/paketo-community_pip/pip/lib/python3.8/site-packages
```

Reply:
Done in https://github.com/paketo-community/pip/pull/109/commits/070220415f6afc0f21003f863718fbaf9e64b549